pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderToken.java

```diff
@@ -458,4 +485,64 @@ public String getHeader(String name) {
             return super.getHeader(name);
         }
     }
+
+    @Slf4j
+    private static final class JwkResolver implements SigningKeyResolver {
+        @Setter
+        @Getter
+        private static class JwksData {
+            private List<Map<String, Object>> keys;
+        }
+
+        private final Map<String, Jwk> jwks;
+
+        public JwkResolver(String configValue) {
+            try {
+                byte[] bytes = AuthTokenUtils.readKeyFromUrl(configValue);
+                ObjectMapper objectMapper = ObjectMapperFactory.create();
+                JwksData data = objectMapper.reader().readValue(bytes, JwksData.class);
+                if (data == null || data.getKeys() == null || data.getKeys().isEmpty()) {
+                    log.warn("No keys in " + data);
+                    jwks = Collections.emptyMap();
+                    return;
+                }
+                jwks = new LinkedHashMap<>();
+                data.getKeys().forEach((n) -> {
+                    Jwk jwk = Jwk.fromValues(n);
+                    jwks.put(jwk.getId(), jwk);
+                });
+                if (log.isDebugEnabled()) {
+                    log.info("jwks: {}", jwks);
+                }
+            } catch (IOException e) {
+                log.error("Failed to get jwks from {}", configValue, e);
+                throw new IllegalArgumentException(e);
+            }
+        }
+
+        private Key get(String keyId) {
```

Comment:
we have an existing key setup in Pulsar creating our JWTs. We dont set a kid though.
could supporting these type of keys and a JWKS file at the same time! be accommodated!? feels like could be a common enough use case for current pulsar users wishing to "migrate" to JWKS for their key mgmt. 

My thoughts on our migration strategy would be put the new/rotated keys in place with the JWKS file, and verify JWTs using this config and / or if needed - the existing key / config. We eventually "turn off" the single key config - and just verify jwts using JWKS. This is critical to avoid disruption of our pulsar clients.  

Thinking through a potential implementation, it seems tricky. One idea would be; again allow both the existing config & new JWKS config be set together. Each of these configs could map to a new interface : impl - and the `AuthenticationProviderToken` will check the incoming JWTs against all these provided impls  

Some initial info anyways, and will see what ur thoughts are @nodece / team

Reply:
And 99% sure a follow on issue / PR makes sense, if u agree thats best

Reply:
Good idea!

> could supporting these type of keys and a JWKS file at the same time!

If the kid is null, and you only have one key in the JWKS file, we can use this key to verify your token.

What do you think?

Reply:
> If the kid is null, and you only have one key in the JWKS file, we can use this key to verify your token.
While that would be correct - to support migrating from a single non-kid key to kid keys.. the JWKS file would need to store both. This general approach is perhaps cleaner / less confusing though than what i proposed. Specifically, the existing, single key config and this new JWKS config can remain separate (and u set one or the other).     

So to support the migration use case, we could introduce a new optional field into the `JwkResolver` - lets calls it `legacyKey` for now. This field would be set when the `JwkResolver` is spinning through the file, and if it sees an entry with no kid, it sets the `legacyKey` field only. Finally, we just update the Key get(String keyId) to return this `legacyKey` field if the incoming JWT has no kid.

Reply:
> If the kid is null, and you only have one key in the JWKS file, we can use this key to verify your token.
> the JWKS file would need to store both.

Sorry, this kid should be your token id.

You only set up one key in the JWKS, which must have the kid. Regardless of whether your token contains an id, you can obtain the correct key, so like:

JWKS hashmap:
```
kid1 -> JWK1
```

| your token id | your JWKS | 
| --- | --- |
| null | JWK1 |
| kid1 | JWK1 |
| kid2 | JWK1 |

> So to support the migration use case we could introduce a new optional field into the JwkResolver

This is more complex rather than my way.

Once all client uses the token with an id, you can add more keys to your JWKS.

Reply:
My idea updated, ping @damienburke

Reply:
Thanks @nodece

I dont think I am putting all the pieces together in your last comment. In your use case table, you are always resolving to JWK1. In particular confused as to how kid2 would resolve to JWK1. Perhaps if u can explain why this happens?  

I think what would really help, if you could state the migration use case(s) your suggested logic supports? Just to clarify our particular migration requirements - we must be able to support verifying JWTs with our existing, non-kid key AND future kid keys - at the same time. (which is supported by my logic).

Not every migration might need to do this - but we have a large pulsar deployment and this "cross-over" is crucial, and the only path for us.

Reply:
> In particular confused as to how kid2 would resolve to JWK1. Perhaps if u can explain why this happens?

You only have one key, not two keys, so you only get the JW1 to verify the token with id kid2. This is a special case.

This special case can support your requirement: > we must be able to support verifying JWTs with our existing, non-kid key AND future kid keys - at the same time.

Does it support configuring two keys to meet this requirement?

No. Please consider which key should be used to verify your token when your token ID is empty.

Reply:
When your token without id, should I use first JWK to verify your token?

Reply:
> When your token without id, should I use first JWK to verify your token?
> Or we need to add the default id to the pulsar config, when your id is null, the pulsar will use this default id to find your JWK.

The "default id" approach sounds cleaner and probably also easier to document.     
Also @nodece - ill be offline until Sunday GMT nite, but will get to any open q's / comments then.

Reply:
Ok, I will open a new PR to do that.